### PR TITLE
fix: require upload ownership before deleting a PPOM file

### PR DESCRIPTION
**Pull request by Alexia-Soare**
Any anonymous visitor who knew or guessed the name of another shopper's in-progress upload could delete it, because `ppom_delete_file` authorised the request with nothing but a nonce that is identical for every logged-out visitor and is handed out publicly at `/wp-json/ppom/v1/nonces/file/`. Deleting now requires having uploaded the file. Reported in [Codeinwp/ppom-pro issue 745](https://github.com/Codeinwp/ppom-pro/issues/745).

The same endpoint pair had two further problems found while testing that fix, both included here because they live in the code this PR already touches.

## What changed

- **Delete guard** — `ppom_delete_file` refuses to remove a file the caller did not upload. Before, a valid nonce plus a file name was enough to delete anyone's upload. An ownership failure reuses the existing nonce-failure wording, so the endpoint does not reveal whether a given file exists.

- **Upload ownership** — each completed upload is recorded against the visitor's WooCommerce session, which is the only thing tying a file in the shared `ppom_files/` directory to the shopper who uploaded it. Recording merges against stored state rather than the snapshot loaded at request start, so an upload that finishes while another is still running is not dropped. The record is removed once the file is deleted, so the list no longer grows on every upload-then-remove cycle.

- **Upload validation** — an upload naming a field that does not exist on the product, or one that exists but cannot upload (a text or select field), is rejected and its file removed. Before, both were finalized, recorded as owned and answered with a preview, leaving a stored file no form could reference. Only `file` and `cropper` fields get an uploader on the frontend; the accepted list runs through a `ppom_upload_capable_field_types` filter so an addon driving this endpoint with its own type can opt in.

- **Preview renderer** — `uploaded_file_preview()` no longer reads `$settings['type']`. `get_field_meta_by_dataname()` returns an empty string on a miss, and indexing it raised a `TypeError` on PHP 8, so an upload with an unknown `product_id` or `data_name` answered 500 after writing the file to disk. That variable and `$data_name` were both unused.

> [!NOTE]
> Nothing changes on the client: `js/file-upload.js` sends exactly what it sent before and the upload response shape is unchanged, so no third-party caller of these endpoints needs updating. Two intentional limits are worth knowing while reviewing. Recording ownership narrows but does not close the concurrent-upload race — uploads that read the session at the same instant still overwrite each other, which needs ownership to stop living in a shared session row, tracked in [Codeinwp/ppom-pro issue 747](https://github.com/Codeinwp/ppom-pro/issues/747). And the upload validation filters malformed and automated traffic rather than acting as a boundary, since a caller can read a real file field's `product_id` and `data_name` off any product page.

## Upload and delete flow

```mermaid
flowchart LR
    U[Upload request] --> UN{Nonce valid?}
    UN -- No --> X[Reject]
    UN -- Yes --> F{New:<br/>field exists and<br/>accepts uploads?}:::added
    F -- No --> R[New: remove file,<br/>return error]:::added
    F -- Yes --> S[Store file] --> REC[New: record against<br/>visitor session]:::added
    REC --> P[Return preview]

    D[Delete request] --> DN{Nonce valid?}
    DN -- No --> X
    DN -- Yes --> O{New:<br/>caller uploaded<br/>this file?}:::added
    O -- No --> X
    O -- Yes --> DEL[Delete file,<br/>drop record]

    classDef added fill:#1a7f37,color:#fff,stroke:#116329,stroke-width:3px
```

## QA

1. **Set up a product with a file field.** Go to **WP Admin → WooCommerce → PPOM Fields** (`/wp-admin/admin.php?page=ppom`; the menu needs the `manage_options` capability, so use an Administrator account) and create a field group containing a **File Upload** field, noting its **Data Name**. Then go to **WP Admin → Products**, open any product, select that group in the **Select PPOM Meta** box in the right-hand sidebar, and click **Update**.

   **Expect:** the product's public page renders the file upload field.

2. **Reported vulnerability.** In a normal browser window, open that product page and upload a file. With DevTools → Network open, find the `admin-ajax.php` request for `ppom_upload_file` and copy `file_name` from its JSON response. Then, from a terminal with no cookies:

   ```bash
   NONCE=$(curl -s "https://SITE/?rest_route=/ppom/v1/nonces/file/" | jq -r .ppom_file_delete_nonce)
   curl -s "https://SITE/wp-admin/admin-ajax.php?action=ppom_delete_file" \
     -d "ppom_nonce=$NONCE" --data-urlencode "file_name=THE_FILE_NAME"
   ```

   **Expect:** `Verification failed for file: …`, and the file is still in `wp-content/uploads/ppom_files/`.

3. **Owner still in control.** Back in that browser window, click **Delete** on the file's preview.

   **Expect:** the file leaves the form and is gone from `wp-content/uploads/ppom_files/`.

4. **Guest who signs in at checkout.** While logged out, upload a file on the product page, add the product to the cart, and sign in to a customer account during checkout. Return to the product page and remove that upload.

   **Expect:** the removal succeeds. Signing in migrates the guest session, and it must not orphan an upload made moments earlier.

5. **Rejected uploads leave nothing behind.** Post an upload naming a field that does not exist, then one naming the **Data Name** of a text field on the same product:

   ```bash
   UP=$(curl -s "https://SITE/?rest_route=/ppom/v1/nonces/file/" | jq -r .ppom_file_upload_nonce)
   curl -s "https://SITE/wp-admin/admin-ajax.php" -F "action=ppom_upload_file" \
     -F "ppom_nonce=$UP" -F "name=a.txt" -F "product_id=PRODUCT_ID" \
     -F "data_name=nope" -F "chunk=0" -F "chunks=1" -F "file=@a.txt"
   ```

   **Expect:** a JSON error rather than a preview or a 500, nothing new in `wp-content/uploads/ppom_files/`, and no `Cannot access offset of type string on string` entry in `wp-content/debug.log`.

### CODE REVIEW

**Review by Soare-Robert-Daniel — APPROVED** [review]

**Review by Soare-Robert-Daniel — APPROVED** [review]

File: src/Files/Handler.php [hunk 1]
```diff
@@ -427,6 +434,59 @@ public static function upload_file() {
...
+		// WooCommerce discards the session on shutdown. Only the stock handler
+		// exposes this; a replacement handler is left to its own persistence.
+		if ( $session instanceof \WC_Session_Handler ) {
+			$session->set_customer_session_cookie( true );
+		}
+
+		$owned   = (array) $session->get( self::OWNED_FILES_KEY, array() );
+		$owned[] = $file_name;
+
+		$session->set( self::OWNED_FILES_KEY, array_values( array_unique( $owned ) ) );
```
[the remark this answers is not in the document]
Comment (Alexia-Soare) [reply src/Files/Handler.php]: Real, and deliberately not fixed here.

Verified: `js/file-upload.js:565` creates one uploader per field, so two can finish concurrently. The variant in your follow-up is the likelier one — a brand-new guest's two concurrent uploads each mint a customer ID and each send `Set-Cookie`, the browser keeps the last, so one upload is recorded against a discarded session. No read-modify-write window needed.

Both reproduced, and worse than my first wording suggested. Three parallel cookieless uploads produced three separate session rows with one file each. On an established session, one sequential upload plus four parallel ones left 2 of 5 names — three lost their record. So the loss is up to N-1 of N concurrent uploads, not an occasional single file.

Two things bound it. The race can only deny a legitimate owner, never authorise an attacker, so this PR's fix holds. And because `die( 0 )` returns 200, `js/file-upload.js` treats the refusal as success and drops the item from the form — the shopper sees nothing, so the symptom is silent orphans rather than a visible error.

An atomic or locked update inside PPOM wouldn't fix either case: `WC_Session_Handler` rewrites the entire session row on shutdown (the cart has the same property), and in the guest case the two requests aren't writing the same row at all.

Your second suggestion is what closes it — a per-upload signed capability verified on delete, dropping the session dependency entirely, at the cost of a client change.

Unbounded list growth is fixed here in 7a27c36.

**Comment by Soare-Robert-Daniel** [comment]
@Alexia-Soare let's explore the low-confidence finding from https://github.com/Codeinwp/woocommerce-product-addon/pull/693#pullrequestreview-4797335651

**Comment by Alexia-Soare** [comment]
@Soare-Robert-Daniel Explored — the finding is accurate, but the delta is smaller than it reads and I'd change the remedy.

Confirmed on this branch: `product_id=99999&data_name=nope` returns 200 with a usable `file_name`.

Pre-fix the same request also wrote the file to disk and *then* 500'd. I re-added the dead line to check: 500, and the orphan still landed. So the write already happened — what's new is that the caller learns the name and nothing is logged.

That's also why validating the field wouldn't remove the abuse surface: a real `product_id`/`data_name` is readable off any product page with a file field, and guests have to be able to upload before checkout. What it does buy is rejecting malformed/bot requests, so fewer orphans — which currently never get swept (#695).

On the specific suggestion, restricting to file/cropper would break `image` and `fancycropper`, plus Pro types using the same uploader. Existence-only is the safer check.

Recommend a separate PR: it needs coverage across those field types, and this one is scoped to the reported delete issue.

**Comment by Alexia-Soare** [comment]
Correcting point 4 of my previous comment — it was wrong.

Only `file` and `cropper` attach to the shared uploader (`js/file-upload.js:338`). `image` is a selection field with no upload. In Pro nothing posts to `ppom_upload_file`, and FancyCropper uses its own REST route (`ppom/v1/save-cropped-images`, `FancyCropper.php:206`). A file/cropper allowlist would break neither codebase.

The error was mine: I read the `case 'cropper'/'fancycropper'/'image'` switch in `update_converted_option_keys()` (`Callbacks.php:1129`) as upload-related. It is option-key normalisation.

The separate-PR recommendation stands, but on different grounds — risk from third-party addons registering types through the same uploader, not breakage of the bundled ones. Points 1-3 are unchanged.

Also stale now: #695 has merged, so the orphan-sweep caveat applies to released versions rather than `development`.

**Review by Soare-Robert-Daniel — APPROVED** [review]

**Comment by kushh23** [comment]
@Soare-Robert-Daniel  should we wait here for others prs to be merged? I see https://github.com/Codeinwp/woocommerce-product-addon/pull/699 and https://github.com/Codeinwp/woocommerce-product-addon/pull/701 wants to merge in this branch, so are they part of this fix?

**Comment by Soare-Robert-Daniel** [comment]
> @Soare-Robert-Daniel should we wait here for others prs to be merged? I see #699 and #701 wants to merge in this branch, so are they part of this fix?

The rest of the PRs are some things that we find while working in #693. If you can check them, it will be merge together.
